## It will check the VENNset class and methods
library(systemPipeR)
skip_on_bioc()
skip_on_ci()
skip("Message")

library(systemPipeRdata)

genWorkenvir("varseq", mydirname = file.path(tempdir(), "varseq"))
setwd(file.path(tempdir(), "varseq"))

test_that("check_varseq", {
    ## Create paths
    targetspath <- system.file("extdata", "targets.txt", package="systemPipeR")
    WF <- loadWorkflow(targets=targetspath, wf_file="hisat2/hisat2-mapping-se.cwl", 
                       input_file="hisat2/hisat2-mapping-se.yml", 
                       dir_path=system.file("extdata/cwl", package="systemPipeR"))
    WF <- renderWF(WF, inputvars=c(FileName="_FASTQ_PATH1_", SampleName="_SampleName_"))
    expect_s4_class(WF, "SYSargs2")
    args <- runCommandline(WF, make_bam = TRUE, dir = FALSE)
    
    outpaths <- subsetWF(args , slot="output", subset=1, index=1)[1:4]
    
    expect_true(all(file.exists(outpaths)))
    
    writeTargetsout(x=WF, file="targets_bam.txt", step=1, new_col=c("FileName"), 
                    new_col_output_index=c(1), overwrite = TRUE)
    
    system("java -jar CreateSequenceDictionary.jar R=./data/tair10.fasta O=./data/tair10.dict")
    # system("java -jar /opt/picard/1.81/CreateSequenceDictionary.jar R=./data/tair10.fasta O=./data/tair10.dict")
    args <- systemArgs(sysma="gatk.param", mytargets="targets_bam.txt")
    resources <- list(walltime="20:00:00", nodes=paste0("1:ppn=", 1), memory="10gb")
    reg <- clusterRun(args, conffile=".BatchJobs.R", template="torque.tmpl", Njobs=18, runid="01",
                      resourceList=resources)
    writeTargetsout(x=args, file="targets_gatk.txt")
    
    ## Variant calling with BCFtools
    ## The following runs the variant calling with BCFtools. This step requires in
    ## the current working directory the parameter file sambcf.param and the 
    ## bash script sambcf_run.sh.
    args <- systemArgs(sysma="sambcf.param", mytargets="targets_bam.txt")
    resources <- list(walltime="20:00:00", nodes=paste0("1:ppn=", 1), memory="10gb")
    reg <- clusterRun(args, conffile=".BatchJobs.R", template="torque.tmpl", Njobs=18, runid="01",
                      resourceList=resources)
    writeTargetsout(x=args, file="targets_sambcf.txt")
    
    ## Filtering of VCF files generated by GATK
    args <- systemArgs(sysma="filter_gatk.param", mytargets="targets_gatk.txt")
    filter <- "totalDepth(vr) >= 2 & (altDepth(vr) / totalDepth(vr) >= 0.8) & rowSums(softFilterMatrix(vr))==4"
    # filter <- "totalDepth(vr) >= 20 & (altDepth(vr) / totalDepth(vr) >= 0.8) & rowSums(softFilterMatrix(vr))==6"
    filterVars(args, filter, varcaller="gatk", organism="A. thaliana")
    writeTargetsout(x=args, file="targets_gatk_filtered.txt")
    
    ## Filtering of VCF files generated by BCFtools
    args <- systemArgs(sysma="filter_sambcf.param", mytargets="targets_sambcf.txt")
    filter <- "rowSums(vr) >= 2 & (rowSums(vr[,3:4])/rowSums(vr[,1:4]) >= 0.8)"
    # filter <- "rowSums(vr) >= 20 & (rowSums(vr[,3:4])/rowSums(vr[,1:4]) >= 0.8)"
    filterVars(args, filter, varcaller="bcftools", organism="A. thaliana")
    writeTargetsout(x=args, file="targets_sambcf_filtered.txt")
    
    ## Annotate filtered variants from GATK
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_gatk_filtered.txt")
    txdb <- loadDb("./data/tair10.sqlite")
    fa <- FaFile(systemPipeR::reference(args))
    variantReport(args=args, txdb=txdb, fa=fa, organism="A. thaliana")
    
    ## Annotate filtered variants from BCFtools
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
    txdb <- loadDb("./data/tair10.sqlite")
    fa <- FaFile(systemPipeR::reference(args))
    variantReport(args=args, txdb=txdb, fa=fa, organism="A. thaliana")
    
    ## Combine results from GATK
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_gatk_filtered.txt")
    combineDF <- combineVarReports(args, filtercol=c(Consequence="nonsynonymous"))
    write.table(combineDF, "./results/combineDF_nonsyn_gatk.xls", quote=FALSE, row.names=FALSE, sep="\t")
    
    ## Combine results from BCFtools
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
    combineDF <- combineVarReports(args, filtercol=c(Consequence="nonsynonymous"))
    write.table(combineDF, "./results/combineDF_nonsyn_sambcf.xls", quote=FALSE, row.names=FALSE, sep="\t")
    
    ## Summary for GATK
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_gatk_filtered.txt")
    write.table(varSummary(args), "./results/variantStats_gatk.xls", quote=FALSE, col.names = NA, sep="\t")
    
    ## Summary for BCFtools
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
    write.table(varSummary(args), "./results/variantStats_sambcf.xls", quote=FALSE, col.names = NA, sep="\t")
    
    ## Venn diagram of variants
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_gatk_filtered.txt")
    varlist <- sapply(names(outpaths(args))[1:4], function(x) as.character(read.delim(outpaths(args)[x])$VARID))
    vennset_gatk <- overLapper(varlist, type="vennsets")
    args <- systemArgs(sysma="annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
    varlist <- sapply(names(outpaths(args))[1:4], function(x) as.character(read.delim(outpaths(args)[x])$VARID))
    vennset_bcf <- overLapper(varlist, type="vennsets")
    vennPlot(list(vennset_gatk, vennset_bcf), mymain="", mysub="GATK: red; BCFtools: blue", colmode=2, ccol=c("blue", "red"))
    })
